[TOMEE-4354] Implicitly create ContextService when needed

jungm opened the pull request:
Resolves [TOMEE-4354](https://issues.apache.org/jira/browse/TOMEE-4354)

Makes the following configuration work again on TomEE 10 (note that Context is omitted):
```
  <Resource id="Mes" type="ManagedExecutorService">
    Core = 5
    Max = 25
    KeepAlive = 5 s
    Queue = 15
    ThreadFactory = org.apache.openejb.threads.impl.ManagedThreadFactoryImpl
    Lazy = true
  </Resource>
```

Also ensures that §3.1.4.3, §3.2.4.3, §3.4.4.3 and §3.3.4.3 of [Concurrency 3.0 spec](https://jakarta.ee/specifications/concurrency/3.0/jakarta-concurrency-spec-3.0) are fulfilled

rmannibucau commented:
maybe just change the default of `openejb.environment.default` instead of the code which prevent to disable it (main case)

jungm commented:
These services were added unconditionally before as well until https://github.com/apache/tomee/commit/2c2fa2d95a68ded5ef52526b27838915730ff339 to fix ConfigurationFactoryTest, I reverted this again and fixed the test differently

rzo1 commented:
Think the original intention of the flag was to prevent creartion of JMS stuff (if one looks into the history)

rmannibucau commented:
Not really, this 9 years old behavior intent is to not create default resources when it is not needed (in all apps except TCK ;)) so think it is sane to revert back to the original behavior and just use the toggle for tests/tcks.

My 2cts

rzo1 commented:
@rmannibucau I see. Just looked in 8.x and it was like that before, so I agree, that we might just want to change the default and people not needing it, can still disable.

rmannibucau commented:
Well there are a lot of things there. If you don't care of the detail just read the last sentence ;).

> It is not about the TCK/tests, it is more or less about the requirement to have a context service available to have an managed executor service.

Hmm I understand "it is 100% about TCK" - cause requirements are just about them for me and embedded defaults/coded behavior about usability/users.

I didn't check recently but when it was added (in 7?) I also added a `Lazy` support for resources but it is way different than this default which disables undesired resources by default (sane default for all cases from embedded daemon to prod ear) and it was still impacting negatively the startup (tests)).
So from my perspective - I still support it but use it way less today - it is wrong to keep it by default until bundled (it is ok to set the toggle to something else in the .zip/.tar.gz since people are unlikely to use it if you follow me.

> Or would you say, the solution for people needing to configure a custom managed executor service is: configure the toggle and be happy about it? (so actually a doc update, which mention it?)

Personally I think we shouldn't enable any default resource but ultimately people willing a tck compatible instance could set it up (this is actually what is done, a ton of flags are needed to respect the spec requirements and nobody cares).
Or we just provide a default system.properties in distro which is compliant (people rarely use it I think if they install tomee properly with base/home).

So I would enphasize the doc, probably explain a bit more the reasoning and why being explicit there is important and explain that we handle default resources - it is openejb default behavior - as soon as a resource of the same type matches or is known it works (tests of openejb-core use that for ex).
Note the name is not important there until you enable exact matching (linking) - another toggle nobody uses probably.

The issue specifically with thread pool: there is no default correct thread pool in java. In a pure world where everything is reactive you are either a worker thread (computing intensive) or IO thread (latency friendly but low need) so from there you already can't get a single thread pool or banalised tasks as in EE concurrency. Now in Java we are still with some blocking tasks which are yet another kind of tasks. When you mix it all with a default executor service algorithm - mainly queue based - you can interlock and you can't - as a container - configure it right for any application, so default never works except in rare test cases but it hides the fact you have to configure it for prod (often multiple instances).

Side note: tomee keeps a default auto resource creation normally (in `AutoConfig`) from service providers so it should just work smoothly - as tests show.

The issue is more an implementation of the factory - wrapper of the actual executor instance - which does a lookup assuming all resources are bound (https://github.com/apache/tomee/blob/main/container/openejb-core/src/main/java/org/apache/openejb/resource/thread/ManagedExecutorServiceImplFactory.java#L74).

This one is quite trivial to fix, instead of failling is `context` is empty, null or missing in JNDI, just do a `new ContextServiceImpl` (maybe with a lazy logger info log line to still catch configuration errors with logs - easy to do using from application resources which have an offset).

jungm commented:
I was thinking about two ways to fix this, either fix the factory so it just works (like you suggested Romain) or to make sure there is a ContextService bound in JNDI

Do you think we should still have to make sure to bind the ContextService (so not just `new ContextServiceImpl()` and call it a day) in the ManagedExecutorService factory? Otherwise there would be a bunch of implicitly created ContextService instances if one configures more than one MES/SMES

jungm commented:
fair enough, I have next to no experience using JNDI so I guess you are right ;)

Singleton in factory sounds like a good option

jungm commented:
reworked now

rmannibucau commented on container/openejb-core/src/main/java/org/apache/openejb/threads/impl/ContextServiceImplFactory.java:
```diff
@@ -77,6 +86,36 @@ public static ContextServiceImpl newPropagateEverythingContextService() {
+    }
+
+    public static ContextServiceImpl lookupOrDefault(String name) {
+        if (name == null || name.trim().isEmpty()) {
```
thinking out loud: I guess the default case would be ok to use the default singleton without logging so maybe we can have a specific token for that ("tomeeDefault" - very bad name but it should not be something xbean layer handles (`@xxx`, `$...`) nor something too common like `default`, maybe `auto!`).
This way we can get `if (auto) return singleton else if (not set) return logAndSingleton() else lookup()` and get a nominal default case.

I'll let you judge on this one, current flavor is fine for me once it gets the lazy logger.

jungm replied:
I've changed the default ContextService name to `[automatic]` now for SMES/MES which will cause the factory to use the singleton. Please take a look if I've done things as you intended 😉

rmannibucau commented on container/openejb-core/src/test/java/org/apache/openejb/threads/ManagedExecutorServiceNoExplicitContextTest.java:
```diff
@@ -0,0 +1,103 @@
+import static org.junit.Assert.assertThat;
+
+@RunWith(ApplicationComposer.class)
+public class ManagedExecutorServiceNoExplicitContextTest {
```
:+1:

rmannibucau reviewed (changes requested):
very small changes (mainly the lazy logger thing) and LGTM

jungm commented on container/openejb-core/src/main/java/org/apache/openejb/config/BuiltInEnvironmentEntries.java:
```diff
@@ -105,7 +105,6 @@ private void add(final JndiConsumer jndi, final DeploymentModule module, final D
             // Jakarta Concurrency §3.3.4.3
             add(jndi.getResourceEnvRefMap(), new ResourceEnvRef().name("java:comp/DefaultContextService").type(ContextService.class));
 
-
```
done

jungm commented on container/openejb-core/src/main/java/org/apache/openejb/threads/impl/ContextServiceImplFactory.java:
```diff
@@ -29,6 +35,9 @@
 
 public class ContextServiceImplFactory {
 
+    private static final Logger LOGGER = Logger.getInstance(LogCategory.OPENEJB, ContextServiceImplFactory.class);
```
done

jungm commented on container/openejb-core/src/main/java/org/apache/openejb/threads/impl/ContextServiceImplFactory.java:
```diff
@@ -77,6 +86,36 @@ public static ContextServiceImpl newPropagateEverythingContextService() {
     }
 
 
+    public static ContextServiceImpl getOrCreateDefaultSingleton() {
```
done

rmannibucau reviewed (approved):
awesome work :+1:

rzo1 commented:
LGTM, too :) - thanks @rmannibucau for review, feedback and (long) explanation. It is really appreciated!

The CI gave some test failures originating from the change:

```
Testergebnis (20 fehlgeschlagene Tests / +18)
org.apache.openejb.arquillian.ManagedExecutorServiceGetPrincipalInTaskTest.run
org.apache.openejb.config.BuiltInEnvironmentEntriesTest.testRed
org.apache.openejb.config.BuiltInEnvironmentEntriesTest.testBlue
org.apache.openejb.threads.CompletableFutureTest.testSupplyAsyncAndCompleteExceptionally
org.apache.openejb.threads.CompletableFutureTest.testSupplyAsync
org.apache.openejb.threads.CompletableFutureTest.testRunAsync
org.apache.openejb.threads.InjectionTest.checkUserTransactionIsAccessible
org.apache.openejb.threads.InjectionTest.checkInjections
org.apache.openejb.threads.ManagedExecutorServiceTest.checkCdiContext
org.apache.openejb.threads.ManagedExecutorServiceTest.runnable
org.apache.openejb.threads.ManagedExecutorServiceTest.checkEjbContext
org.apache.openejb.threads.SecurityPropagationTest.checkItIsTrue
org.apache.openejb.server.cxf.rs.SuspendedTest.run
org.superbiz.executor.ManagedScheduledServiceTest.singleFixedDelayTaskWithException
org.superbiz.executor.ManagedScheduledServiceTest.singleFixedDelayTask
org.superbiz.executor.ManagedScheduledServiceTest.periodicFixedDelayTaskWithException
org.superbiz.executor.ManagedScheduledServiceTest.periodicFixedDelayTask
org.superbiz.executor.ManagedServiceTest.managedInvocationTest
org.superbiz.executor.ManagedServiceTest.managedInvocationTestWithException
org.superbiz.executor.ManagedServiceTest.managedInvocationTestWithTimeout
```

Full logs here: https://ci-builds.apache.org/job/Tomee/view/tomee-10.x/job/pull-request-manual/113/#showFailuresLink 

FYI: @jungm

jungm commented:
New build is green except for a test I can't get to fail locally: https://ci-builds.apache.org/job/Tomee/view/tomee-10.x/job/pull-request-manual/116/
